### PR TITLE
Ordino 2.0 Working Branch

### PR DESCRIPTION
**Pull request by dvzacharycutler**
This PR as well as the PRs linked below are intended to be an up to date version of the current progress of Ordino 2.0 

https://github.com/datavisyn/tdp_core/pull/602
https://github.com/datavisyn/reprovisyn/pull/185
https://github.com/datavisyn/ordino_reprovisyn/pull/10

Current progress in this repo:
Converting Ordino to be fully react, with a redux state.
Creating the basis for our new UI, with a chevron filmstrip menu and a dashboard-like workbench. Views in a workbench can be resized and moved. Transitions from one workbench to another.

### CODE REVIEW

File: src/app/workbench/useLoadViewPlugin.ts [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+    const {status, value: viewPlugin} = useAsync(loadView, []);
+
+    const setRef = React.useCallback(async (ref: HTMLElement | null) => {
+        // Create a new one if there is a ref
+        if (ref && status === 'success') {
+
+            const idType = workbenchIndex === 0 ? 'Start' : ordino.workbenches[workbenchIndex - 1].entityId;
+
+            const selection = {idtype: new IDType(idType, viewId, '', true), range: workbenchIndex === 0 ? Range.none() : Range.list(ordino.workbenches[workbenchIndex - 1].selections)};
```
Comment (thinkh) [inline src/app/workbench/useLoadViewPlugin.ts:27]: Refactoring suggestion:

```suggestion
            const previousWorkbench = ordino.workbenches[workbenchIndex - 1];
            
            const idType = (previousWorkbench) ? previousWorkbench.entityId : Start;
            const range = (previousWorkbench) ? Range.list(previousWorkbench.selections) : Range.none();
            
            const selection = {idtype: new IDType(idType, viewId, '', true), range};
```

File: src/app/workbench/useLoadViewPlugin.ts [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+            FindViewUtils.findAllViews(selection.idtype).then((availableViews) => {
+                const filteredViews = availableViews.filter((v) => viewId.endsWith(v.v.itemIDType));
+
+                const context = {graph: null, ref: {value: {data: null}} as any, desc: workbenchIndex === 0 ? view : filteredViews[0].v};
+
+                const i = viewPlugin.factory(context, selection, ref, {});
+                context.ref[`v`] = i;
+
+                console.log('#' + viewId);
```
Comment (thinkh) [inline src/app/workbench/useLoadViewPlugin.ts:49]: Remove?

```suggestion
```

File: src/app/workbench/useLoadViewPlugin.ts [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    const setRef = React.useCallback(async (ref: HTMLElement | null) => {
+        // Create a new one if there is a ref
+        if (ref && status === 'success') {
+
+            const idType = workbenchIndex === 0 ? 'Start' : ordino.workbenches[workbenchIndex - 1].entityId;
+
+            const selection = {idtype: new IDType(idType, viewId, '', true), range: workbenchIndex === 0 ? Range.none() : Range.list(ordino.workbenches[workbenchIndex - 1].selections)};
+
+
+            FindViewUtils.findAllViews(new IDType(viewId, '.*', '', true)).then((availableViews) => {
```
Comment (thinkh) [inline src/app/workbench/useLoadViewPlugin.ts:30]: Add a comment that adresses the difference between this `findAllViews` and the next one in line 41? Why not using `await`?

File: src/app/workbench/useLoadViewPlugin.ts [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+                availableViews.forEach((v) => {
+                    idTargetSet.add(v.v.itemIDType);
+                });
+
+                dispatch(addTransitionOptions({transitionOptions: Array.from(idTargetSet.values()), workbenchIndex}));
+
+            });
+
+            FindViewUtils.findAllViews(selection.idtype).then((availableViews) => {
```
Comment (thinkh) [inline src/app/workbench/useLoadViewPlugin.ts:41]: I don't understand what is going on in this function. Please add a comment.

Why not using `await`?

File: src/app/workbench/useLoadViewPlugin.ts [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+                const filteredViews = availableViews.filter((v) => viewId.endsWith(v.v.itemIDType));
+
+                const context = {graph: null, ref: {value: {data: null}} as any, desc: workbenchIndex === 0 ? view : filteredViews[0].v};
+
+                const i = viewPlugin.factory(context, selection, ref, {});
```
Comment (thinkh) [inline src/app/workbench/useLoadViewPlugin.ts:46]: What is `i`? Does it mean `instance`? Please rename.

File: src/app/workbench/useLoadViewPlugin.ts [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+                ResolveNow.resolveImmediately(i.init(document.getElementById(viewId).querySelector('.view-parameters'), () => null)).then(() => i.setInputSelection(selection));
+                setInstance(i);
+            });
+        }
+    }, [status]);
+
+    React.useEffect(() => {
+        if(instance) {
+            const view: ARankingView = instance as unknown as ARankingView;
```
Comment (thinkh) [inline src/app/workbench/useLoadViewPlugin.ts:59]: This does not seem reliable. Maybe you can check at least if the instance is an ARankingView or `viewType === 'Ranking'`?

File: src/app/workbench/useLoadViewPlugin.ts [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+            const id = IDTypeManager.getInstance().resolveIdType(view.itemIDType.id);
+
+            view.selectionHelper.setGeneralVisSelection({idtype: id, range: Range.list(ordino.workbenches[workbenchIndex].selections)});
+
+        }
+    }, [instance, ordino.workbenches[workbenchIndex].selections]);
+
+    React.useEffect(() => {
+        if(instance) {
+            const view: ARankingView = instance as unknown as ARankingView;
```
Comment (thinkh) [inline src/app/workbench/useLoadViewPlugin.ts:69]: Same here.

File: src/components/breadcrumb/TempChevronJumpButtons.tsx [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+                                    selections: [],
+                                })
+                            );
+                            setTimeout(() => {
+                                dispatch(
+                                    changeFocus({index: ordino.focusViewIndex + 1})
+                                );
+                            }, 0);
+
+                        }}type="button" className="chevronButton btn btn-light btn-sm align-middle m-1" style={{color}}>Jump to {j.name}</button>
```
Comment (thinkh) [inline src/components/breadcrumb/TempChevronJumpButtons.tsx:65]: Missing space + extract lanugage string.


```suggestion
                        }} type="button" className="chevronButton btn btn-light btn-sm align-middle m-1" style={{color}}>Jump to {j.name}</button>
```

File: src/scss/abstracts/_variables.scss [hunk 9]
```diff
@@ -6,6 +6,6 @@ $session-list-max-height: 31em !default; // show 6 items
 $chooser-width: 220px !default;
 $chooser-collapsed-width: 50px !default;
 
-$workbench-width: calc(100vw - #{$chooser-collapsed-width}) !default;
-$workbench-focus-width: calc(90vw - #{$chooser-collapsed-width}) !default;
-$workbench-hide-width: 10vw !default;
+$workbench-width: calc(100vw) !default;
+$workbench-focus-width: calc(85vw) !default;
+$workbench-context-width: 15vw !default;
```
Comment (thinkh) [inline src/scss/abstracts/_variables.scss:11]: Add a comment that these widths must be in sync with the one in the TS file. Please also add a comment to the TS file.

File: src/store/ordinoSlice.ts [hunk 10]
```diff
@@ -150,18 +141,28 @@ const ordinoSlice = createSlice({
...
-      // state.workbenches[action.payload.workbenchIndex].views[action.payload.viewIndex].filters.push(action.payload.newFilter);
+    addFilter(state, action: PayloadAction<{viewId: string, filter: number[]}>) {
+      state.workbenches[state.focusViewIndex].views.find((v) => v.id === action.payload.viewId).filters = action.payload.filter;
     },
     changeFocus(state, action: PayloadAction<{index: number}>) {
       state.focusViewIndex = action.payload.index;
-    }
+    },
+    setWorkbenchData(state, action: PayloadAction<{data: any[]}>) {
+      for(const i of action.payload.data) {
```
Comment (thinkh) [inline src/store/ordinoSlice.ts:154]: Is `i` a row? If yes, please rename.

```suggestion
      for(const row of action.payload.data) {
```

File: src/store/ordinoSlice.ts [hunk 11]
```diff
@@ -150,18 +141,28 @@ const ordinoSlice = createSlice({
...
+        state.workbenches[state.focusViewIndex].data[i._id] = i;
+      }
+    },
+    addScoreColumn(state, action: PayloadAction<{columnName: string, data: any}>) {
+      for(const i of action.payload.data) {
```
Comment (thinkh) [inline src/store/ordinoSlice.ts:159]: Is `i` a row? If yes, please rename.

```suggestion
      for(const row of action.payload.data) {
```

File: src/store/storeUtils.ts [hunk 12]
```diff
@@ -0,0 +1,13 @@
+import {IWorkbench} from '.';
+
+export function getAllFilters(workbench: IWorkbench) {
+    const allFilteredIds: Set<number> = new Set<number>();
+    workbench.views.forEach((f) => {
+        f.filters.forEach((id) => {
```
Comment (thinkh) [inline src/store/storeUtils.ts]: Possibly rename

```suggestion
    workbench.views.forEach((view) => {
        view.filters.forEach((id) => {
```

File: tsconfig.json [hunk 13]
```diff
@@ -24,6 +24,6 @@
     "src/**/*.ts",
     "src/**/*.tsx",
     "tsd.d.ts"
-  ],
+, "../ordino_reprovisyn/src/dataSlice.ts"  ],
```
Comment (thinkh) [inline tsconfig.json]: Why is this necessary here? The `ordino` should not reference `ordino_reprovisyn` since it is and cannot be a depencency in the package.json.

File: src/scss/components/Workbench.scss [hunk 14]
```diff
@@ -1,87 +1,122 @@
...
-    background-position: 50%;
-  }
+.view-title {
+  font-size: 16px;
+}
 
-  .workbenchOverlayButton {
-    opacity: 1;
-  }
+.Pane {
```
Comment (thinkh) [inline src/scss/components/Workbench.scss:50]: Why do some CSS classes start with a upper-case letter? Is this a requirement from a dependent library? Otherwise please rename these CSS class names.

**Review by thinkh — CHANGES_REQUESTED** [review]
I made some suggestions and minor refactorings.

In general the `useLoadViewPlugin` seems fuzzy to me and not very reliable. I don't know if this might change with loading react view. But we should add at least some comments to make this code more understandable.

Other than that the plotly.js dependency and tsconfig.json should be reviewed.

File: src/app/Workbench.tsx [hunk 15]
```diff
@@ -59,42 +48,44 @@ export function Workbench({workbench, type = EWorkbenchType.PREVIOUS, onScrollTo
     const onReplaceView = React.useCallback((view: IViewPluginDesc, viewIndex: number) => {
         dispatch(
             replaceWorkbench({workbenchIndex: viewIndex, newWorkbench: {
-                viewDirection: 'vertical',
-                views: [{}],
-                id: view.id,
+                viewDirection: 'horizontal',
+                views: [{viewType: 'Ranking', filters: [], index: 0}],
```
[the remark this answers is not in the document]
Comment (dvzacharycutler) [reply src/app/Workbench.tsx:52]: viewType is getting changed to viewId immediately, so no need. Added enum for direction

File: src/scss/components/Workbench.scss (hunk 14 above)
Comment (dvzacharycutler) [reply src/scss/components/Workbench.scss:50]: Yes, this is me overriding classes that come from the react-split library

**Comment by dvzacharycutler** [comment]
For now I did not touch the useLoadViewPlugin, as I agree with Holger it is not reliable and the entire thing will be removed once we shift to using react views, which is coming in the next PR

**Review by dvchristianbors — APPROVED** [review]
